packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -3,50 +3,197 @@
+///     focus information and actions to its descendant widgets.
+///   * [FocusScope], a widget that manages a [FocusScopeNode] and provides
+///     access to scope information and actions to its descendant widgets.
+///   * [FocusManager], a singleton that manages the focus and distributes key
+///     events to focused nodes.
+///   * [FocusTraversalPolicy], an object used to determine how to move the
+///     focus to other nodes.
+///   * [DefaultFocusTraversal], a widget used to configure the default focus
+///     traversal policy for a widget subtree.
+class FocusNode with DiagnosticableTreeMixin, ChangeNotifier {
+  /// Creates a Focus node.
+  ///
+  /// All parameters are optional.
+  FocusNode({
+    String debugLabel,
+    this.onKey,
+  }) {
+    // Set it this way so that it does nothing on release builds.
+    this.debugLabel = debugLabel;
+  }
+
+  /// The context that was supplied to [reparentIfNeeded].
+  ///
+  /// This is typically the context in the build method for the widget that is
+  /// being focused, as it is used to determine the bounds of the widget.
+  BuildContext get context => _context;
+  BuildContext _context;
+
+  /// Called if this focus node receives a key event while focused (i.e. when
+  /// [hasFocus] returns true).
+  ///
+  /// It doesn't need to be the primary focus to receive events: it will also
+  /// receive key events if it is the ancestor of the node with the primary
+  /// focus (i.e. the one that returns true for [hasPrimaryFocus]), unless the
+  /// primary focus or one of its closer ancestors has returned true from its
+  /// [onKey], indicating that it handled the event.
+  ///
+  /// {@macro flutter.widgets.focus_manager.focus.keyEvents}
+  FocusOnKeyCallback onKey;
+
   FocusManager _manager;
   bool _hasKeyboardToken = false;
 
-  /// Whether this node has the overall focus.
+  /// Returns the parent node for this object.
+  FocusNode get parent => _parent;
+  FocusNode _parent;
+
+  /// An iterator over the children of this node.
+  Iterable<FocusNode> get children => _children;
+  final List<FocusNode> _children = <FocusNode>[];
+
+  /// A debug label that is used for diagnostic output.
   ///
-  /// A [FocusNode] has the overall focus when the node is focused in its
-  /// parent [FocusScopeNode] and [FocusScopeNode.isFirstFocus] is true for
+  /// Will always return null in release builds.
+  String get debugLabel => _debugLabel;
+  String _debugLabel;
+  set debugLabel(String value) {
+    assert(() {
+      // Only set the value in debug builds.
+      _debugLabel = value;
+      return true;
+    }());
+  }
+
+  /// An iterator over the hierarchy of children below this one, in depth first
+  /// order.
+  Iterable<FocusNode> get descendants sync* {
+    for (FocusNode child in _children) {
+      yield* child.descendants;
+      yield child;
+    }
+  }
+
+  /// An iterator over the ancestors of this node.
```

Comment:
ditto.

this should say at a minimum what the order will be (furthest ancestor first? parent first?)